flang/lib/Lower/IO.cpp

```diff
@@ -196,6 +196,144 @@ static void makeNextConditionalOn(Fortran::lower::FirOpBuilder &builder,
+///         const Descriptor &descriptor;
+///       };
+///       const char *groupName; // NUL-terminated lower-case
+///       std::size_t items;
+///       const Item *item; // in original declaration order
+///     };
+///
+/// Return the Item array and the NamelistGroup data structure.
+static std::pair<fir::GlobalOp, fir::GlobalOp>
+getNamelistGroup(Fortran::lower::AbstractConverter &converter,
+                 const Fortran::semantics::Symbol &symbol) {
+  auto &builder = converter.getFirOpBuilder();
+  auto mangledGroupName = converter.mangleName(symbol);
+  auto listName = mangledGroupName + ".list";
+  if (auto group = builder.getNamedGlobal(mangledGroupName)) {
+    auto list = builder.getNamedGlobal(listName);
+    assert(list && "missing namelist group list");
+    return {list, group};
+  }
+  const auto &details = symbol.get<Fortran::semantics::NamelistDetails>();
+  auto loc = converter.getCurrentLocation();
+  auto *context = builder.getContext();
+  auto linkOnce = builder.createLinkOnceLinkage();
+  auto idxTy = builder.getIndexType();
+  auto sizeTy = builder.getIntegerType(8 * sizeof(std::size_t));
+  auto charPtrTy = fir::PointerType::get(builder.getIntegerType(8));
+  auto descPtrTy =
+      fir::PointerType::get(fir::BoxType::get(mlir::NoneType::get(context)));
+  auto listTy = fir::SequenceType::get(
+      details.objects().size(),
+      mlir::TupleType::get(context, {charPtrTy, descPtrTy}));
+  auto groupTy = mlir::TupleType::get(
+      context, {charPtrTy, sizeTy, fir::PointerType::get(listTy)});
+  auto stringAddress = [&](const Fortran::semantics::Symbol &symbol,
+                           bool wantResult = false) {
+    auto global =
+        createStringLiteral(builder, loc, symbol.name().ToString() + '\0');
+    return wantResult
+               ? builder.createConvert(loc, charPtrTy, fir::getBase(global))
+               : mlir::Value{};
+  };
+  stringAddress(symbol);
+  for (const Fortran::semantics::Symbol &s : details.objects())
+    stringAddress(s);
+  size_t n;
+  auto nextIdx = [&]() {
+    return builder.createIntegerConstant(loc, idxTy, n++);
+  };
+  auto listFunc = [&](Fortran::lower::FirOpBuilder &builder) {
+    mlir::Value list = builder.create<fir::UndefOp>(loc, listTy);
+    auto zero = builder.createIntegerConstant(loc, idxTy, 0);
+    n = 0;
+    for (const Fortran::semantics::Symbol &s : details.objects()) {
+      llvm::SmallVector<mlir::Value, 2> idx = {nextIdx(), zero};
+      auto name = stringAddress(s, true);
+      list = builder.create<fir::InsertValueOp>(loc, listTy, list, name, idx);
+    }
+    builder.create<fir::HasValueOp>(loc, list);
+  };
+  auto list =
+      builder.createGlobal(loc, listTy, listName, false, listFunc, linkOnce);
+  auto groupFunc = [&](Fortran::lower::FirOpBuilder &builder) {
+    mlir::Value group = builder.create<fir::UndefOp>(loc, groupTy);
+    n = 0;
+    auto groupName = stringAddress(symbol, true);
```

Comment:
Same here, would the following work ?:
```
auto groupName = builder.createConvert(loc, charPtrTy, fir::getBase(createStringLiteral(...));
auto groupFunc = [&](Fortran::lower::FirOpBuilder &builder) {
   ....
  group = builder.create<fir::InsertValueOp>(loc, groupTy, group, groupName,
}
```

Reply:
All namelist component accesses now use `fir::InsertValueOp`.